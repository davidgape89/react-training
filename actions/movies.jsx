import axios from 'axios';

const baseUrl = 'https://reactjs-cdp.herokuapp.com';

export const startMoviesRequest = (query, searchBy) => 
  (dispatch) => axios.get(`${baseUrl}/movies?search=${query}&searchBy=${searchBy}`)
    .then(({data}) => {
      dispatch(addMovies(data.data))
    });

export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  payload: {
    movies
  }
});