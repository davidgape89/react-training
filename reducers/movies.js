export default (state = [], {type, payload}) => {
  switch(type) {
    case 'ADD_MOVIES':
      return payload.movies;
    default:
      return state;
  }
}