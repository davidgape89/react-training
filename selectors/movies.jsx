import { createSelector } from 'reselect';

const getMovies = state => state.movies;

const getSortBy = state => state.filters.sortBy;

const getSortedMovies = createSelector(
  [getMovies, getSortBy],
  (movies, sortBy) => {
    const resultArray = [...movies];

    if (sortBy === 'release_date') {
      return resultArray.sort(
        (movie1, movie2) => {
          const timestamp1 = new Date(movie1.release_date);
          const timestamp2 = new Date(movie2.release_date);
          return timestamp1 < timestamp2 ? 1 : -1;
        },
      );
    }

    return resultArray.sort(
      (movie1, movie2) => (movie1.vote_average < movie2.vote_average ? 1 : -1),
    );
  },
);

export default getSortedMovies;
