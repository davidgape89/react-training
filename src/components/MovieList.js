import React from 'react';
import MoviePoster from './MoviePoster';

export default ({movies, handleMovieSelected}) => (
  <div className="movie-list">
    {movies.map((movie) => (
      <MoviePoster key={movie.id}
                  id={movie.id}
                  title={movie.title}
                  genres={movie.genres}
                  posterUrl={movie.poster_path}
                  releaseDate={movie.release_date}
                  handleMovieSelected={handleMovieSelected} />
    ))}
  </div>
);