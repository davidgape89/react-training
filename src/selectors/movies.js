export default (movies, sortBy) => {
  let resultArray = [...movies];
  if(sortBy === 'release_date') {
    return resultArray.sort(
      (movie1, movie2) => {
        const timestamp1 = new Date(movie1.release_date);
        const timestamp2 = new Date(movie2.release_date);
        return timestamp1 < timestamp2? 1: -1;
      });
  } else {
    return resultArray.sort(
      (movie1, movie2) => {
        return movie1.vote_average < movie2.vote_average? 1: -1;
      }
    );
  }
}