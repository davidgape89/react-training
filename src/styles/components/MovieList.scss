@import '../variables';

.movie-list {
  background-color: $extra-light-grey;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  padding: $big-size;

  .movie-poster {
    margin-bottom: $big-size;
  }
}
