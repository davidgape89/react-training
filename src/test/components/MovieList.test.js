import React from 'react';
import {shallow} from 'enzyme';

import movies from '../../mocks/movie';
import MovieList from '../../components/MovieList';

describe('MovieList -', () => {
  let wrapper, callbackFun;

  beforeEach(() => {
    callbackFun = jest.fn();
    wrapper = shallow(<MovieList movies={movies} 
                                 onPageChange={callbackFun} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the callback when the children component does', () => {
    wrapper.find('MoviePoster')
           .first()
           .prop('onPageChange')();

    expect(callbackFun).toHaveBeenCalled();
  })
});