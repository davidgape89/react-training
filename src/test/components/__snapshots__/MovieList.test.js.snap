// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList - renders correctly 1`] = `
<div
  className="movie-list"
>
  <MoviePoster
    genres={
      Array [
        "Drama",
        "Romance",
      ]
    }
    handleMovieSelected={[MockFunction]}
    id={337167}
    key="337167"
    posterUrl="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
    releaseDate="2018-02-07"
    title="Fifty Shades Freed"
  />
  <MoviePoster
    genres={
      Array [
        "Animation",
        "Adventure",
        "Family",
        "Comedy",
      ]
    }
    handleMovieSelected={[MockFunction]}
    id={269149}
    key="269149"
    posterUrl="https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg"
    releaseDate="2016-02-11"
    title="Zootopia"
  />
  <MoviePoster
    genres={
      Array [
        "Fantasy",
        "Adventure",
        "Science Fiction",
      ]
    }
    handleMovieSelected={[MockFunction]}
    id={181808}
    key="181808"
    posterUrl="https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg"
    releaseDate="2017-12-13"
    title="Star Wars: The Last Jedi"
  />
  <MoviePoster
    genres={
      Array [
        "Action",
        "Adventure",
        "Fantasy",
        "Science Fiction",
      ]
    }
    handleMovieSelected={[MockFunction]}
    id={284054}
    key="284054"
    posterUrl="https://image.tmdb.org/t/p/w500/uxzzxijgPIY7slzFvMotPv8wjKA.jpg"
    releaseDate="2018-02-13"
    title="Black Panther"
  />
</div>
`;
